cincinnati/src/plugins/internal/arch_filter.rs

```diff
@@ -0,0 +1,356 @@
+//! This plugin can be used to filter a graph by a architecture.
+//! It reads the requested architecture from the parameters value at key "arch",
+//! and the value must match the regex specified at ARCH_VALIDATION_REGEX_STR
```

Comment:
We only hit this code after we've loaded the graph, right?  I'd rather drop the regexp and just track the set of all architectures seen during graph building.  Then if someone requests an arch we haven't seen we can say "invalid arch {whatever-they-gave-us}, try one of these {arches-we've-seen}".  If they give us a version string, we may even be able to filter that set and just return arch choices for that particular version, falling back to the set of all seen architectures if we don't recognize the version.

Reply:
> We only hit this code after we've loaded the graph, right?

The plugin receives the graph as part of its input when being executed in the context of a single request. I consider the plugins to be able to have certain requirements towards their input. I don't see the necessity to introduce a state here for keeping a history of all architectures, as the graph contains that state already. If a user requests an architecture which validates but isn't in the graph, they'll get an empty graph, which is fine IMO.

Reply:
> If a user requests an architecture which validates but isn't in the graph, they'll get an empty graph, which is fine IMO.

That's fair.  But then why place validation constraints on the architecture string at all?

Reply:
So that the user knows the difference between requesting an invalid architecture and an empty graph. The graph might be empty for a multitude of reasons which leaves the user guessing.

Reply:
> So that the user knows the difference between requesting an invalid architecture and an empty graph.

What is the difference?  If you want to return an empty graph if the user asks for `arch=nevergonnahappen`, I don't see why you wouldn't also return an empty graph (and no error) for users who request `arch=never+gonna+happen`.

Reply:
How about `arch=amd-64` vs `arch=amd64`? I'm thinking of the case where the user provides an arch which will eventually emit a non-empty graph for the request which he's making now, but he has a typo in the string which we could tell them about.

Reply:
> I'm thinking of the case where the user provides an arch which will eventually emit a non-empty graph for the request which he's making now, but he has a typo in the string which we could tell them about.

I'm fine with that to.  But a generic regexp approach is only going to catch so many of those.  For example, it's not going to catch `nevergonnahappen` unless folks are manually maintaining the regexp to list expected arches.  And what would you put in a list of expected arches besides the arches you have in your graph?  So I see too paths:

a. The "user knows what they're doing" path, where we just accept whatever they gave for `arch` and return an empty graph with no error.
b. The "our graph has the set of arches" path, where we error out in this plugin with "unrecognised arch: `${whatever}`" if arch-filtering by the given `arch` empties the graph.

Trying to sit somewhere between those two seems like manual effort for questionable benefit.

Reply:
I think (a) and (b) can be combined. We definitely know and will tell the user if they provided an architecture which is incorrect by means of the underlying technology used, namely SemVer build annotations, which we can inform the user about. We could additionally tell them that the graph doesn't contain their syntactically correct architecture as an error, I'd be fine with that. But I wouldn't track the list of architectures over time. Rather I'd compile it for each request from the input graph of the plugin.

Reply:
> We definitely know and will tell the user if they provided an architecture which is incorrect by means of the underlying technology used, namely SemVer build annotations, which we can inform the user about.

I'm having trouble parsing this.  Can you give a concrete example?

> We could additionally tell them that the graph doesn't contain their syntactically correct architecture as an error, I'd be fine with that.

I don't think there's a need to separate out "syntactically correct"; just say "error!  No matches for your provided `arch=${ARCH}`".

> But I wouldn't track the list of architectures over time. Rather I'd compile it for each request from the input graph of the plugin.

So we agree?  Can we drop `ARCH_VALIDATION_REGEX_STR`?

Reply:
> I'm having trouble parsing this. Can you give a concrete example?

If the user provides an architecture which contains an invalid character according to the [SemVer specification](https://semver.org/#spec-item-10https://semver.org/#spec-item-10) we can let them know. This may be caused due to a simple typo, e.g. "amd64.". This is different from an architecture name which is valid but has no releases in the graph.

> I don't think there's a need to separate out "syntactically correct"; just say "error! No matches for your provided `arch=${ARCH}`".

We don't *need* to do anything :smile: I'd do it to give the possibly most precise error message to the user. I don't understand where your resistance towards distinguishing these two cases is coming from :man_shrugging:

> So we agree? Can we drop `ARCH_VALIDATION_REGEX_STR`?

Not really. We could have the regex check *and* the non-existing arch check. The latter would naturally take precedence as it's cheaper than looking through the graph. But if you really feel so strongly about it I can also remove the regex and only leave the non-existing arch check in.